save-agent/src/nativeMain/kotlin/org/cqfn/save/agent/SaveAgent.kt

```diff
@@ -0,0 +1,155 @@
+                    }
+                }
+) {
+    /**
+     * The current [AgentState] of this agent
+     */
+    val state = AtomicReference(AgentState.IDLE)
+    private val id = uuid4().toString()
+    private val isStopped = atomic(false)
+    private var saveProcessJob: Job? = null
+
+    /**
+     * @return Unit
+     */
+    suspend fun start() = coroutineScope {
+        println("Starting agent")
+        val heartbeatsJob = launch { startHeartbeats() }
+        heartbeatsJob.join()
+        println("Heartbeats job is done")
+    }
+
+    /**
+     * Shutdown the Agent
+     */
+    fun stop() {
+        println("Stopping agent")
+        httpClient.close()
+        if (saveProcessJob?.isActive == true) {
+            saveProcessJob?.cancel()
+        }
+        isStopped.getAndSet(true)
+    }
+
+    @Suppress("WHEN_WITHOUT_ELSE")  // when with sealed class
+    private suspend fun startHeartbeats() = coroutineScope {
+        println("Scheduling heartbeats")
+        while (isStopped.value.not()) {
+            val response = runCatching { sendHeartbeat() }
+            if (response.isSuccess) {
+                when (response.getOrNull()) {
+                    is NewJobResponse -> maybeStartSaveProcess()
+                    TerminatingResponse -> {
+                        println("Terminating the agent because termination signal has been received")
+                        stop()
+                    }
+                    EmptyResponse -> Unit  // do nothing
+                }
+            } else {
+                println("Exception during heartbeat: ${response.exceptionOrNull()?.message}")
+            }
+            // todo: start waiting after request was sent, not after response?
+            println("Waiting for ${config.heartbeatIntervalMillis} sec")
+            delay(config.heartbeatIntervalMillis)
+        }
+    }
+
+    private suspend fun maybeStartSaveProcess() = coroutineScope {
+        if (saveProcessJob?.isCompleted == false) {
+            println("Shouldn't start new process when there is the previous running")
+        } else {
+            saveProcessJob = launch {
+                // new job received from Orchestrator, spawning SAVE CLI process
+                startSaveProcess()
+            }
+        }
+    }
+
+    /**
+     * @return Unit
+     */
+    internal suspend fun startSaveProcess() = coroutineScope {
+        // blocking execution of OS process
+        // val code = saveAgent.runSave(emptyList())
+        state.value = AgentState.FINISHED
+        // todo: read data from files here
+        sendExecutionData(ExecutionData(emptyList()))
+        state.value = AgentState.IDLE
+    }
+
+    private fun runSave(cliArgs: List<String>) = platform.posix.system("./save ${cliArgs.joinToString(" ")}")
+
+    /**
+     * @return a [HeartbeatResponse] from Orchestrator
+     */
+    internal suspend fun sendHeartbeat(): HeartbeatResponse {
```

Comment:
create a data class with the values that will be posted to orchestrator and pass it as argument